File: dailyqs/1657_words.py
class Solution:
    def closeStrings(self, word1: str, word2: str) -> bool:
        l1 = {}
        l2 = {}
        if len(word1) != len(word2):
            return False
        for c1, c2 in zip(word1, word2):
            l1[c1] = l1.get(c1, 0) + 1
            l2[c2] = l2.get(c2, 0) + 1
        
        # both words must have the same set of letters
        for c in l1.keys():
            if c not in l2:
                return False

        occs1 = sorted(l1.values())
        occs2 = sorted(l2.values())
        return occs1 == occs2
    
testCases = [
    ["abc", "bca"],
    ["a", "aa"],
    ["cabbba", "abbccc"],
    ["asa", "ssx"]
]

for words in testCases:
    s = Solution()
    print(f'{words[0]=} and {words[1]=}: {s.closeStrings(words[0], words[1])}')

File: dailyqs/1657_words.js
// @ts-check

/**
 * @param {string} word1
 * @param {string} word2
 * @return {boolean}
 */
function closeStrings(word1, word2) {
    if (word1.length !== word2.length) {
        return false;
    }

    /** @type Map<string, number> */
    const l1 = new Map()
    /** @type Map<string, number> */
    const l2 = new Map();

    for (var i = 0; i < word1.length; i++) {
        const c1 = word1[i], c2 = word2[i];
        l1.set(c1, (l1.get(c1) ?? 0) + 1);
        l2.set(c2, (l2.get(c2) ?? 0) + 1);
    }
    if (l1.size !== l2.size) {
        // number of unique letters differs between two words
        return false;
    }

    // we need same letters in both words
    for (const c of l1.keys()) {
        if (!l2.has(c)) {
            return false;
        }
    }

    // sort occurences and compare, we need same occurences in 
    // both strings after sorting by number of occurences
    var occs1 = Array.from(l1.values()).sort((a, b) => a - b);
    var occs2 = Array.from(l2.values()).sort((a, b) => a - b);
    return occs1.every((value, index) => value === occs2[index]);
};

const testCases = [
    ["abc", "bca"],
    ["a", "aa"],
    ["cabbba", "abbccc"],
    ["asa", "ssx"]
];

for (const words of testCases) {
    console.log(words[0] + " and " + words[1] + ": " + closeStrings(words[0], words[1]));
}
